perses/rjmc/topology_proposal.py

```diff
@@ -2729,20 +3258,168 @@ def preserves_rings(match, current_mol, proposed_mol):
         return True
 
     @staticmethod
-    def rank_degenerate_maps(old_mol, new_mol, matches):
+    def _create_pattern_to_target_map(current_mol, proposed_mol, match, matching_criterion = 'index'):
+        """
+        Create a dict of {pattern_atom: target_atom}
+
+        Arguments
+        ---------
+        current_mol : openeye.oechem.oemol object
+        proposed_mol : openeye.oechem.oemol object
+        match : oechem.OEMCSSearch.Match iter
+            entry in oechem.OEMCSSearch.Match object
+        matching_criterion : str, default 'index'
+            whether the pattern to target map is chosen based on atom indices or names (which should be uniquely defined)
+            allowables: ['index', 'name']
+
+        Returns
+        -------
+        pattern_to_target_map : dict
+            {pattern_atom: target_atom}
+        """
+        if matching_criterion == 'index':
+            pattern_atoms = { atom.GetIdx() : atom for atom in current_mol.GetAtoms() }
+            target_atoms = { atom.GetIdx() : atom for atom in proposed_mol.GetAtoms() }
+            pattern_to_target_map = { pattern_atoms[matchpair.pattern.GetIdx()] : target_atoms[matchpair.target.GetIdx()] for matchpair in match.GetAtoms() }
+        elif matching_criterion == 'name':
+            pattern_atoms = {atom.GetName(): atom for atom in current_mol.GetAtoms()}
+            target_atoms = {atom.GetName(): atom for atom in proposed_mol.GetAtoms()}
+            pattern_to_target_map = {pattern_atoms[matchpair.pattern.GetName()]: target_atoms[matchpair.target.GetName()] for matchpair in match.GetAtoms()}
+        else:
+            raise Exception(f"matching criterion {matching_criterion} is not currently supported")
+        return pattern_to_target_map
+
+
+    @staticmethod
+    def check_molecule_name_uniqueness(molecule):
+        """
+        check that the oemol has unique name identifiers
+
+        Arguments
+        ---------
+        molecule : openeye.oechem.oemol
+            molecule to check
+
+        Returns
+        -------
+        validated : bool
+            whether the names are unique identifiers
+        """
+        if not len(set([atom.GetName() for atom in molecule.GetAtoms()])) == len([atom.GetName() for atom in molecule.GetAtoms()]):
+            _logger.warning(f"\t\t\tthe molecule atoms are not uniquely named")
+            return False
+        else:
+            return True
+
+
+    @staticmethod
+    def preserve_chirality(current_mol, proposed_mol, new_to_old_atom_map):
```

Comment:
first pass to allow for proper handling of chirality; this is not particularly robust, but gets the job done for a decent number of applications we are interested in at the moment; read docstrings and comments to see the details of chiral center handling